src/terminal/input/terminalInput.cpp

```diff
@@ -513,209 +172,466 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
+    // the RightAlt bit on the final key state. If we don't clear that, it
+    // can be misinterpreted as an Alt+AltGr key combination.
+    const auto rightAltDown = key.bKeyDown && key.wVirtualKeyCode == VK_MENU && WI_IsFlagSet(key.dwControlKeyState, ENHANCED_KEY);
+    WI_ClearFlagIf(_lastControlKeyState, RIGHT_ALT_PRESSED, WI_IsFlagSet(pressedKeyState, RIGHT_ALT_PRESSED) && !rightAltDown);
+    // We also take this opportunity to record the time at which the LeftCtrl
+    // and RightAlt keys are pressed. This is needed to determine whether the
+    // Ctrl key was pressed by the user, or fabricated by an AltGr key press.
+    if (key.bKeyDown)
     {
-        return _makeCharOutput(keyEvent.uChar.UnicodeChar);
+        if (WI_IsFlagSet(pressedKeyState, LEFT_CTRL_PRESSED))
+        {
+            _lastLeftCtrlTime = GetTickCount64();
+        }
+        if (WI_IsFlagSet(pressedKeyState, RIGHT_ALT_PRESSED))
+        {
+            _lastRightAltTime = GetTickCount64();
+        }
     }
-
-    return MakeUnhandled();
+    return _lastControlKeyState;
 }
 
-TerminalInput::OutputType TerminalInput::HandleFocus(const bool focused) const
+// Returns a simplified representation of the keyboard state, based on the most
+// recent key press and associated control key state (which is all we need for
+// our ToUnicodeEx queries). This is a substitute for the GetKeyboardState API,
+// which can't be used when serving as a conpty host.
+std::array<byte, 256> TerminalInput::_getKeyboardState(const WORD virtualKeyCode, const DWORD controlKeyState) const
 {
-    if (!_inputMode.test(Mode::FocusEvent))
+    auto keyState = std::array<byte, 256>{};
+    if (virtualKeyCode < keyState.size())
     {
-        return MakeUnhandled();
+        keyState.at(virtualKeyCode) = 0x80;
     }
+    keyState.at(VK_LCONTROL) = WI_IsFlagSet(controlKeyState, LEFT_CTRL_PRESSED) ? 0x80 : 0;
+    keyState.at(VK_RCONTROL) = WI_IsFlagSet(controlKeyState, RIGHT_CTRL_PRESSED) ? 0x80 : 0;
+    keyState.at(VK_CONTROL) = keyState.at(VK_LCONTROL) | keyState.at(VK_RCONTROL);
+    keyState.at(VK_LMENU) = WI_IsFlagSet(controlKeyState, LEFT_ALT_PRESSED) ? 0x80 : 0;
+    keyState.at(VK_RMENU) = WI_IsFlagSet(controlKeyState, RIGHT_ALT_PRESSED) ? 0x80 : 0;
+    keyState.at(VK_MENU) = keyState.at(VK_LMENU) | keyState.at(VK_RMENU);
+    keyState.at(VK_SHIFT) = keyState.at(VK_LSHIFT) = WI_IsFlagSet(controlKeyState, SHIFT_PRESSED) ? 0x80 : 0;
+    keyState.at(VK_CAPITAL) = WI_IsFlagSet(controlKeyState, CAPSLOCK_ON);
+    return keyState;
+}
 
-    return MakeOutput(focused ? L"\x1b[I" : L"\x1b[O");
+wchar_t TerminalInput::_makeCtrlChar(const wchar_t ch)
+{
+    if (ch >= L'@' && ch <= L'~')
+    {
+        return ch & 0b11111;
+    }
+    if (ch == L'/')
+    {
+        return 0x1F;
+    }
+    if (ch == L'?')
+    {
+        return 0x7F;
+    }
+    if (ch >= L'2' && ch <= L'8')
+    {
+        constexpr auto numericCtrls = std::array<wchar_t, 8>{ 0, 27, 28, 29, 30, 31, 127 };

```

Comment:
this has 7 things in it

Reply:
```suggestion
        constexpr auto numericCtrls = std::array<wchar_t, 7>{ 0, 27, 28, 29, 30, 31, 127 };
```

watch, i'll take this suggestion and explode the world

Reply:
Oh, I didn't realize there was another non-static `std::array`! Oh well... We can fix that whenever. I'm sure we have more of those.